embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/ApiClientImpl.kt

```diff
@@ -0,0 +1,158 @@
+package io.embrace.android.embracesdk.comms.api
+
+import io.embrace.android.embracesdk.comms.api.ApiClient.Companion.NO_HTTP_RESPONSE
+import io.embrace.android.embracesdk.comms.api.ApiClient.Companion.defaultTimeoutMs
+import io.embrace.android.embracesdk.logging.InternalEmbraceLogger
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.io.InputStream
+import java.io.InputStreamReader
+import java.net.HttpURLConnection
+import java.util.zip.GZIPOutputStream
+
+/**
+ * Client for calling the Embrace API. This service handles all calls to the Embrace API.
+ *
+ * Sessions can be sent to either the production or development endpoint. The development
+ * endpoint shows sessions on the 'integration testing' screen within the dashboard, whereas
+ * the production endpoint sends sessions to 'recent sessions'.
+ *
+ * The development endpoint is only used if the build is a debug build, and if integration
+ * testing is enabled when calling [Embrace.start()].
+ */
+internal class ApiClientImpl(
+    private val logger: InternalEmbraceLogger
+) : ApiClient {
+
+    override fun executeGet(request: ApiRequest): ApiResponse<String> {
+        var connection: EmbraceConnection? = null
+
+        try {
+            connection = request.toConnection()
+            setTimeouts(connection)
+            connection.connect()
+            return executeHttpRequest(connection)
+        } catch (ex: Throwable) {
+            throw IllegalStateException(ex.localizedMessage ?: "", ex)
+        } finally {
+            runCatching {
+                connection?.inputStream?.close()
+            }
+        }
+    }
+
+    override fun executePost(request: ApiRequest, payloadToCompress: ByteArray): ApiResponse<String> =
+        executeRawPost(request, gzip(payloadToCompress))
+
+    /**
+     * Posts a payload according to the ApiRequest parameter. The payload will not be gzip compressed.
+     */
+    private fun executeRawPost(request: ApiRequest, payload: ByteArray?): ApiResponse<String> {
+        logger.logDeveloper("ApiClient", request.httpMethod.toString() + " " + request.url)
+        logger.logDeveloper("ApiClient", "Request details: $request")
+
+        var connection: EmbraceConnection? = null
+        return try {
+            connection = request.toConnection()
+            setTimeouts(connection)
+            if (payload != null) {
+                logger.logDeveloper("ApiClient", "Payload size: " + payload.size)
+                connection.outputStream?.write(payload)
```

Comment:
future: thoughts on serializing JSON (or cached file) directly to the request outputstream?

Reply:
We are caching the payload as bytes to enable retries. @lucaslabari is actually going to do the gzipping of those bytes earlier, but I wonder if caching a zipped (or unzipped, depending on size) payload and doing the serialization here would make more sense.

Serializing is so expensive and if we can defer that and still don't blow up disk usage, that might be worth trying!

Let me profile this later.